File: csharp/BrainfuckArray.cs
public class BrainfuckArray : Benchmark
{
    private string _programText = "";
    private string _warmupText = "";
    private uint _result;

    public override uint Checksum => _result;
    public override string TypeName => "Brainfuck::Array";

    public BrainfuckArray()
    {
        _programText = Helper.Config_s(TypeName, "program");
        _warmupText = Helper.Config_s(TypeName, "warmup_program");
    }

    private struct Tape
    {
        private byte[] _tape;
        private int _pos;

        public Tape()
        {
            _tape = new byte[30000];
            _pos = 0;
        }

        public byte Get() => _tape[_pos];
        public void Inc() => _tape[_pos] = (byte)(_tape[_pos] + 1);
        public void Dec() => _tape[_pos] = (byte)(_tape[_pos] - 1);

        public void Advance()
        {
            _pos += 1;
            if (_pos >= _tape.Length)
            {
                Array.Resize(ref _tape, _tape.Length + 1);
            }
        }

        public void Devance()
        {
            if (_pos > 0) _pos -= 1;
        }
    }

    private class Program
    {
        private List<byte> _commands;
        private int[] _jumps;

        public Program(string text)
        {
            _commands = new List<byte>(text.Length);
            foreach (char c in text)
            {
                switch (c)
                {
                    case '[':
                    case ']':
                    case '<':
                    case '>':
                    case '+':
                    case '-':
                    case ',':
                    case '.':
                        _commands.Add((byte)c);
                        break;
                }
            }

            _jumps = new int[_commands.Count];
            int[] stack = new int[_commands.Count];
            int sp = 0;

            for (int i = 0; i < _commands.Count; i++)
            {
                if (_commands[i] == '[')
                {
                    stack[sp++] = i;
                }
                else if (_commands[i] == ']' && sp > 0)
                {
                    int start = stack[--sp];
                    _jumps[start] = i;
                    _jumps[i] = start;
                }
            }
        }

        public uint Run()
        {
            var tape = new Tape();
            int pc = 0;
            var commands = _commands;
            var jumps = _jumps;
            uint result = 0;

            while (pc < commands.Count)
            {
                byte cmd = commands[pc];

                switch (cmd)
                {
                    case (byte)'+': tape.Inc(); break;
                    case (byte)'-': tape.Dec(); break;
                    case (byte)'>': tape.Advance(); break;
                    case (byte)'<': tape.Devance(); break;
                    case (byte)'[':
                        if (tape.Get() == 0)
                        {
                            pc = jumps[pc];
                            continue;
                        }
                        break;
                    case (byte)']':
                        if (tape.Get() != 0)
                        {
                            pc = jumps[pc];
                            continue;
                        }
                        break;
                    case (byte)'.':
                        result = unchecked((result << 2) + tape.Get());
                        break;
                }
                pc++;
            }

            return result;
        }
    }

    private uint RunProgram(string text)
    {
        return new Program(text).Run();
    }

    public override void Warmup()
    {
        long prepareIters = WarmupIterations;
        for (long i = 0; i < prepareIters; i++)
        {
            RunProgram(_warmupText);
        }
    }

    public override void Run(long IterationId)
    {
        _result = unchecked(_result + RunProgram(_programText));
    }
}

File: zig/src/brainfuck_array.zig
const std = @import("std");
const Benchmark = @import("benchmark.zig").Benchmark;
const Helper = @import("helper.zig").Helper;

pub const BrainfuckArray = struct {
    allocator: std.mem.Allocator,
    helper: *Helper,
    program_text: []const u8,
    warmup_text: []const u8,
    result_val: u32,

    const Tape = struct {
        tape: []u8,
        pos: usize,

        pub fn init(allocator: std.mem.Allocator) !Tape {
            const tape = try allocator.alloc(u8, 30000);
            @memset(tape, 0);
            return Tape{ .tape = tape, .pos = 0 };
        }

        pub fn deinit(self: *Tape, allocator: std.mem.Allocator) void {
            allocator.free(self.tape);
        }

        pub fn get(self: *const Tape) u8 {
            return self.tape[self.pos];
        }

        pub fn inc(self: *Tape) void {
            self.tape[self.pos] +%= 1;
        }

        pub fn dec(self: *Tape) void {
            self.tape[self.pos] -%= 1;
        }

        pub fn advance(self: *Tape, allocator: std.mem.Allocator) !void {
            self.pos += 1;
            if (self.pos >= self.tape.len) {
                const new_len = self.tape.len + 1;
                const new_tape = try allocator.realloc(self.tape, new_len);
                new_tape[self.tape.len] = 0;
                self.tape = new_tape;
            }
        }

        pub fn devance(self: *Tape) void {
            if (self.pos > 0) {
                self.pos -= 1;
            }
        }
    };

    const Program = struct {
        commands: []u8,
        jumps: []usize,
        allocator: std.mem.Allocator,

        pub fn init(allocator: std.mem.Allocator, text: []const u8) !Program {
            var commands_list = std.ArrayList(u8).initCapacity(allocator, text.len) catch return error.OutOfMemory;
            defer commands_list.deinit(allocator);

            for (text) |c| {
                if (std.mem.indexOfScalar(u8, "[]<>+-,.", c) != null) {
                    commands_list.appendAssumeCapacity(c);
                }
            }

            const commands = try allocator.alloc(u8, commands_list.items.len);
            @memcpy(commands, commands_list.items);

            const jumps = try allocator.alloc(usize, commands.len);
            @memset(jumps, 0);

            var stack = std.ArrayList(usize).initCapacity(allocator, commands.len / 2) catch return error.OutOfMemory;
            defer stack.deinit(allocator);

            for (commands, 0..) |cmd, idx| {
                switch (cmd) {
                    '[' => {
                        stack.appendAssumeCapacity(idx);
                    },
                    ']' => {
                        if (stack.pop()) |start| {
                            jumps[start] = idx;
                            jumps[idx] = start;
                        }
                    },
                    else => {},
                }
            }

            return Program{
                .commands = commands,
                .jumps = jumps,
                .allocator = allocator,
            };
        }

        pub fn deinit(self: *Program) void {
            self.allocator.free(self.commands);
            self.allocator.free(self.jumps);
        }

        pub fn run(self: *Program) !u32 {
            var result: u32 = 0;
            var tape = try Tape.init(self.allocator);
            defer tape.deinit(self.allocator);

            var pc: usize = 0;
            const commands = self.commands;
            const jumps = self.jumps;

            while (pc < commands.len) {
                const cmd = commands[pc];
                switch (cmd) {
                    '+' => tape.inc(),
                    '-' => tape.dec(),
                    '>' => tape.advance(self.allocator) catch return error.OutOfMemory,
                    '<' => tape.devance(),
                    '[' => {
                        if (tape.get() == 0) {
                            pc = jumps[pc];
                            continue;
                        }
                    },
                    ']' => {
                        if (tape.get() != 0) {
                            pc = jumps[pc];
                            continue;
                        }
                    },
                    '.' => {
                        result = (result << 2) +% tape.get();
                    },
                    else => {},
                }
                pc += 1;
            }

            return result;
        }
    };

    const vtable = Benchmark.VTable{
        .run = runImpl,
        .checksum = resultImpl,
        .deinit = deinitImpl,
        .warmup = warmupImpl,
    };

    pub fn init(allocator: std.mem.Allocator, helper: *Helper) !*BrainfuckArray {
        const program_text = helper.config_s("Brainfuck::Array", "program");
        const warmup_text = helper.config_s("Brainfuck::Array", "warmup_program");

        const self = try allocator.create(BrainfuckArray);
        errdefer allocator.destroy(self);

        self.* = BrainfuckArray{
            .allocator = allocator,
            .helper = helper,
            .program_text = try allocator.dupe(u8, program_text),
            .warmup_text = try allocator.dupe(u8, warmup_text),
            .result_val = 0,
        };
        return self;
    }

    pub fn deinit(self: *BrainfuckArray) void {
        self.allocator.free(self.program_text);
        self.allocator.free(self.warmup_text);
        self.allocator.destroy(self);
    }

    pub fn asBenchmark(self: *BrainfuckArray) Benchmark {
        return Benchmark.init(self, &vtable, self.helper, "Brainfuck::Array");
    }

    fn runImpl(ptr: *anyopaque, iteration_id: i64) void {
        _ = iteration_id;
        const self: *BrainfuckArray = @ptrCast(@alignCast(ptr));

        var program = Program.init(self.allocator, self.program_text) catch return;
        defer program.deinit();

        const result = program.run() catch return;
        self.result_val +%= result;
    }

    fn resultImpl(ptr: *anyopaque) u32 {
        const self: *BrainfuckArray = @ptrCast(@alignCast(ptr));
        return self.result_val;
    }

    fn deinitImpl(ptr: *anyopaque) void {
        const self: *BrainfuckArray = @ptrCast(@alignCast(ptr));
        self.deinit();
    }

    fn warmupImpl(ptr: *anyopaque) void {
        const self: *BrainfuckArray = @ptrCast(@alignCast(ptr));
        if (self.warmup_text.len == 0) return;

        var program = Program.init(self.allocator, self.warmup_text) catch return;
        defer program.deinit();
        _ = program.run() catch return;
    }
};
